ikarus/utils/observer/observer.hh

```diff
@@ -138,169 +77,84 @@ public:
+  void unSubscribeAll(std::shared_ptr<IObserver<IObservable>> observer);
   /**
    * \brief Notify observers about a specific message type.
    * \param message The message type to notify about.
+   * \param state The state information needed for logging.
    */
-  void notify(MessageType message);
-
-  /**
-   * \brief Notify observers about a specific message type with a floating-point value.
-   * \tparam ScalarType The type of the floating-point value.
-   * \param message The message type to notify about.
-   * \param val The floating-point value to be sent with the notification.
-   */
-  template <std::floating_point ScalarType>
-  void notify(MessageType message, ScalarType val);
-
-  /**
-   * \brief Notify observers about a specific message type with an integer value.
-   * \param message The message type to notify about.
-   * \param val The integer value to be sent with the notification.
-   */
-  void notify(MessageType message, int val);
-
-  /**
-   * \brief Notify observers about a specific message type with a string value.
-   * \param message The message type to notify about.
-   * \param val The string value to be sent with the notification.
-   */
-  void notify(MessageType message, const std::string& val);
-
-  /**
-   * \brief Notify observers about a specific message type with an integer and a double value.
-   * \param message The message type to notify about.
-   * \param val1 The integer value to be sent with the notification.
-   * \param val2 The double value to be sent with the notification.
-   */
-  void notify(MessageType message, int val1, double val2);
-
-  /**
-   * \brief Notify observers about a specific message type with an integer value and a string value.
-   * \param message The message type to notify about.
-   * \param val1 The integer value to be sent with the notification.
-   * \param val2 The string value to be sent with the notification.
-   */
-  void notify(MessageType message, int val1, const std::string& val2);
-
-  /**
-   * \brief Notify observers about a specific message type with an Eigen::VectorX.
-   * \tparam ScalarType The type of the elements in the Eigen::VectorX.
-   * \param message The message type to notify about.
-   * \param vec The Eigen::VectorX to be sent with the notification.
-   */
-  template <std::floating_point ScalarType>
-  void notify(MessageType message, Eigen::VectorX<ScalarType> vec);
+  void notify(MessageType message, const StateType& state);
 
 private:
-  using ObserverVector = std::vector<std::shared_ptr<IObserver<MessageType>>>;
+  using ObserverVector = std::vector<std::shared_ptr<IObserver<IObservable>>>;
   using ObserverMap    = std::map<MessageType, ObserverVector>;
   ObserverMap observers_;
   std::vector<MessageType> messages_;
 };
 
-template <typename MT>
-void IObservable<MT>::subscribe(MT message, std::shared_ptr<IObserver<MT>> observer) {
+template <typename MT, typename ST>
+requires(Concepts::Observable<MT, ST>)
```

Comment:
Here may be we can discuss where these functions are to be defined to avoid repeating this `concept` check, or perhaps there is a better solution?